regression (1.0.4): `{.push exportc.}` wrongly affects generic instantiations, causing codegen errors

regression: `{.push exportc.}` doesn't work anymore in some cases, causing `reset` to be cgen'd without mangling, which causes c compilation error as soon as multiple types are invovled. EDIT: root cause seems to be that exportc applies recursively in some cases (when it involves generics that have at least 1 pragma)

### Example
after upgrading to latest nim, my project failed; after a painful reduction it boils down to this:
```nim
  {.push exportc.}
  proc foo1() =
    var s1 = "bar"
    reset(s1)
    var s2 = @[1]
    reset(s2)
  {.pop.}
```

### Current Output
```
/Users/timothee/git_clone/nim/timn/build/nimcache/@mt0854.nim.c:77:17: error: conflicting types for 'reset'
N_NIMCALL(void, reset)(tySequence__qwqHTkRvwhrRyENtudHQ7g** obj);
                ^
/Users/timothee/git_clone/nim/timn/build/nimcache/@mt0854.nim.c:75:17: note: previous declaration is here
N_NIMCALL(void, reset)(NimStringDesc** obj);
```

### Expected Output
works

### Possible Solution
* `{.push exportc.}` should not be recursive and only apply to top-level
* alternatively if push is meant to be recursive, we need a non-recursive variant, eg:
`{.pushNodepth exportc.}`

### Additional Information
D20191227T162616

* the snippet above was working in version 1.0.4; EDIT: whether it's a regression or not is debatable given that the underlying root cause
* fails in latest devel b67dea7a33d24cd54afa8bd37677d066f4be0da1
* `proc foo1()  {.exportc.}` instead of `{.push exportc.}` does not have that bug

after further analysis the root cause is that `{.push exportc.}` doesn't interact well with `{.noSideEffect.}`:

```nim
  proc timnFoo[T](obj: T) {.noSideEffect.} = discard # BUG
  # proc timnFoo[T](obj: T) = discard # works

  {.push exportc.}
  proc foo1() =
    var s1 = "bar"
    timnFoo(s1)
    var s2 = @[1]
    timnFoo(s2)
  {.pop.}
```

gives
```
@mt0854.nim.c:80:17: error: conflicting types for 'timnFoo'
N_NIMCALL(void, timnFoo)(tySequence__qwqHTkRvwhrRyENtudHQ7g* obj);
```
and timnFoo is not mangled, which is a bug;

whereas if we remove `noSideEffect` it works, and the generic is correctly mangled:
```
N_LIB_PRIVATE N_NIMCALL(void, timnFoo__vtgkL9aHp9bVX9bQn9cY0l4vLA)(tySequence__qwqHTkRvwhrRyENtudHQ7g* obj);
```

--- a/compiler/pragmas.nim
+++ b/compiler/pragmas.nim
@@ -1168,7 +1168,7 @@ proc implicitPragmas*(c: PContext, sym: PSym, n: PNode,
   if sym != nil and sym.kind != skModule:
     for it in c.optionStack:
       let o = it.otherPragmas
-      if not o.isNil:
+      if not o.isNil and sfFromGeneric notin sym.flags: # see issue #12985
         pushInfoContext(c.config, n.info)
         var i = 0
         while i < o.len:

